### PR TITLE
SCT-641 enable search by emergency ids

### PR DESCRIPTION
**Pull request by LBHTKarki**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/SCT-641

### *What is the problem we're trying to solve*

Currently users have to prefix emergency IDs in person search. This is often problematic because users might not know what type of ID they have i.e. emergency or Mosaic.

### *What changes have we introduced*

Enable users to search person records by giving either emergency id or Mosaic id. Users don't have to provide prefix the emergency ids anymore.

This also removes the logic the were only emergency IDs with NC prefix were supported. This is important so we can support them all in the future.

New implementation will automatically find all matching person records regardless of what id type was used. 

#### _Checklist_

- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Inform users that it is no longer necessary to use the prefix although they can still provide it if they want.

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        [Test]
+        public void IfLimitMoreThanTheMaximumWillUseTheMaximumLimit()
+        {
+            _mockDatabaseGateway.Setup(x => x.GetResidentsBySearchCriteria(0, 100, null, null, null, null, null, null, null))
+                .Returns(new List<ResidentInformation>()).Verifiable();
+
+            _getAllUseCase.Execute(new ResidentQueryParam(), 0, 400);
+
+            _mockDatabaseGateway.Verify();
+        }
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs:73]: I think we could improve the readability and understanding of these tests as I struggled at first. For example, seeing the lone `_mockDatabaseGateway.Verify();` as the expectation was initially confusing. I eventually got it and understood that we're expecting that the database gateway to be called with a limit of 10/100 but because we've stubbed the gateway method to only reply to those arguments so we can just use `Verify()`.

Maybe we can remove the stubbing here as we're already doing it in the `SetUp` and update the expectation `Verify()` to check the arguments to be 10/100?

I also think it would help to extract variables for these numbers e.g. `var maximumLimit = 100` so it's clearer.

Comment (chubberlisk) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs:73]: Possibly silly question but what happens on the boundary when the limit is 10? 😅

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        public void ReturnsTheNextCursor()
+        {
+            var residents = _fixture.CreateMany<ResidentInformation>(10);
+            int idCount = 10;
+            foreach (ResidentInformation resident in residents)
+            {
+                idCount++;
+                resident.MosaicId = idCount.ToString();
+            }
+            var expectedNextCursor = residents.Max(r => r.MosaicId);
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs:85]: Could you explain what we're doing here in the setup? Why do we need to change the `MosaicId`s of the residents?

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+            var listOfMatchingIds = new List<long> { 1, 2 };
+
+            _mockDatabaseGateway.Setup(x => x.GetPersonIdsByEmergencyId(It.IsAny<long>())).Returns(listOfMatchingIds);
+            _mockDatabaseGateway.Setup(x => x.GetPersonsByListOfIds(It.IsAny<List<long>>())).Returns(new List<Person>() { TestHelpers.CreatePerson() });
+
+            _getAllUseCase.Execute(request, cursor: 0, limit: 4);
+
+            _mockDatabaseGateway.Verify(x => x.GetPersonsByListOfIds(listOfMatchingIds));
+        }
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs:183]: What do you think about adding `TestCase`s for different `MosaicId`s? As the code mentions we want to be removing prefixes and other letters.

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs (hunk 1 above)
Comment (LBHTKarki) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs]: I have updated the setup and used constants for the limits. Hopefully it's more clear now. I've also added tests for the boundary values to make it clear what happens in those situations.

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs (hunk 2 above)
Comment (LBHTKarki) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs:84]: When fixture library creates those person objects it uses Ids like MosaicId18e8cb3e-d6d8-4e4b-9165-3ad8413d38a5. That idCount together with the loop resets the IDs to values we use (long), so the next cursor calculation can be performed correctly and reflects the types we use.

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs (hunk 3 above)
Comment (LBHTKarki) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs]: I've added test cases for each

File: SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs (hunk 2 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/Residents/GetAllUseCaseTests.cs:84]: Ahhhhh okay, I see thanks!

**Review by chubberlisk — APPROVED** [review]
Thanks for addressing my comments! 🙇‍♀️
